plugin/code_actions.py

```diff
@@ -7,227 +7,195 @@
-        return [a for a in (actions or []) if a.get('kind') in kinds]
 
-    collector = actions_collector.create_collector(config_name)
-    return (
-        lambda actions: collector(actions_filter(actions)),
-        lambda error: collector([])
-    )
+actions_manager = CodeActionsManager()
 
 
-actions_manager = CodeActionsManager()
+def get_session_kinds(session: Session) -> List[CodeActionKind]:
+    session_kinds = session.get_capability('codeActionProvider.codeActionKinds')  # type: Optional[List[CodeActionKind]]
+    return session_kinds or []
 
 
-def get_matching_kinds(user_actions: Dict[str, bool], session_actions: List[CodeActionKind]) -> List[CodeActionKind]:
+def get_matching_on_save_kinds(
+    user_actions: Dict[str, bool], session_kinds: List[CodeActionKind]
+) -> List[CodeActionKind]:
     """
-    Filters user-enabled or disabled actions so that only ones matching the session actions
-    are returned. Returned actions are those that are enabled and are not overridden by more
-    specific, disabled actions.
+    Filters user-enabled or disabled actions so that only ones matching the session kinds
+    are returned. Returned kinds are those that are enabled and are not overridden by more
+    specific, disabled kinds.
 
-    Filtering only returns actions that exactly match the ones supported by given session.
+    Filtering only returns kinds that exactly match the ones supported by given session.
     If user has enabled a generic action that matches more specific session action
     (for example user's a.b matching session's a.b.c), then the more specific (a.b.c) must be
-    returned as servers must receive only actions that they advertise support for.
+    returned as servers must receive only kinds that they advertise support for.
     """
     matching_kinds = []
-    for session_action in session_actions:
+    for session_kind in session_kinds:
         enabled = False
-        action_parts = cast(str, session_action).split('.')
+        action_parts = session_kind.split('.')
         for i in range(len(action_parts)):
             current_part = '.'.join(action_parts[0:i + 1])
             user_value = user_actions.get(current_part, None)
             if isinstance(user_value, bool):
                 enabled = user_value
         if enabled:
-            matching_kinds.append(session_action)
+            matching_kinds.append(session_kind)
     return matching_kinds
 
 
+def kinds_include_kind(kinds: List[CodeActionKind], kind: Optional[CodeActionKind]) -> bool:
+    """
+    The "kinds" include "kind" if "kind" matches one of the "kinds" exactly or one of the "kinds" is a prefix
+    of the whole "kind" (where prefix must be followed by a dot).
+    """
+    if not kind:
+        return False
+    for kinds_item in kinds:
+        if kinds_item == kind:
+            return True
+        kinds_item_len = len(kinds_item)
+        if kind.startswith(kinds_item) and kind[kinds_item_len] == '.':
```

Comment:
```suggestion
        if len(kind) > kinds_item_len and kind.startswith(kinds_item) and kind[kinds_item_len] == '.':
```

You need this check too, otherwise `kind[kinds_item_len]` could throw an IndexError